storage/block.go

```diff
@@ -57,67 +60,111 @@ func NewDatabaseBlock(start time.Time, encoder m3db.Encoder, opts m3db.DatabaseO
 		opts:     opts,
 		start:    start,
 		encoder:  encoder,
+		stream:   nil,
 		ctxAlloc: ctxAlloc,
 		ctx:      ctxAlloc(),
 		closed:   false,
+		writable: true,
 	}
 }
 
 func (b *dbBlock) StartTime() time.Time {
 	return b.start
 }
 
+func (b *dbBlock) IsSealed() bool {
+	return !b.writable
+}
+
 func (b *dbBlock) Write(timestamp time.Time, value float64, unit xtime.Unit, annotation []byte) error {
 	if b.closed {
 		return errWriteToClosedBlock
 	}
+	if !b.writable {
+		return errWriteToSealedBlock
+	}
 	return b.encoder.Encode(m3db.Datapoint{Timestamp: timestamp, Value: value}, unit, annotation)
 }
 
 func (b *dbBlock) Stream(blocker m3db.Context) (m3db.SegmentReader, error) {
 	if b.closed {
 		return nil, errReadFromClosedBlock
 	}
-	s := b.encoder.Stream()
 	if blocker != nil {
 		b.ctx.DependsOn(blocker)
 	}
+	s := b.stream
+	if b.writable {
+		s = b.encoder.Stream()
+	}
 	return s, nil
 }
 
-// close closes internal context and returns encoder to pool.
-func (b *dbBlock) closeContextAndEncoder() {
+// close closes internal context and returns encoder and stream to pool.
+func (b *dbBlock) close() {
 	// If the context is nil (e.g., when it's just obtained from the pool),
 	// we return immediately.
 	if b.ctx == nil {
 		return
 	}
-	if encoder := b.encoder; encoder != nil {
-		b.ctx.RegisterCloser(encoder.Close)
+	if b.writable {
+		// If the block is not sealed, we need to close the encoder.
+		if encoder := b.encoder; encoder != nil {
+			b.ctx.RegisterCloser(encoder.Close)
+		}
+	} else {
+		// Otherwise, we need to close the stream.
+		if stream := b.stream; stream != nil {
+			closer := func() {
+				if bytesPool := b.opts.GetBytesPool(); bytesPool != nil {
+					segment := stream.Segment()
+					stream.Reset(m3db.Segment{})
+					bytesPool.Put(segment.Head)
+					bytesPool.Put(segment.Tail)
+				}
+				stream.Close()
+			}
+			b.ctx.RegisterCloser(closer)
+		}
 	}
+
 	b.ctx.Close()
 	b.ctx = nil
 	b.encoder = nil
+	b.stream = nil
 }
 
 // Reset resets the block start time and the encoder.
 func (b *dbBlock) Reset(startTime time.Time, encoder m3db.Encoder) {
-	b.closeContextAndEncoder()
+	b.close()
 	b.start = startTime
 	b.encoder = encoder
+	b.stream = nil
 	b.ctx = b.ctxAlloc()
 	b.closed = false
+	b.writable = true
 }
 
 func (b *dbBlock) Close() {
 	if b.closed {
 		return
 	}
-	b.closeContextAndEncoder()
+	b.closed = true
+	b.close()
 	if pool := b.opts.GetDatabaseBlockPool(); pool != nil {
 		pool.Put(b)
 	}
-	b.closed = true
+}
+
+func (b *dbBlock) Seal() {
+	if b.closed || !b.writable {
+		return
+	}
+	b.writable = false
+	b.stream = b.encoder.Stream()
+	b.encoder.ResetSetData(b.start, nil, false)
```

Comment:
Don't need to `ResetSetData` to nil before closing.  Users will Reset when they pull it off the pool with their data.